File: dictionary/csharp/data-processing.cs
// Arithmetic mean of a list of doubles.
public static double AverageOf(List<double> values)
{
    return values.Count == 0 ? 0.0 : values.Average();
}

// Min-max normalization mapping values into [0, 1].
public static List<double> NormalizeValues(List<double> values)
{
    double min = values.Min();
    double max = values.Max();
    if (Math.Abs(max - min) < double.Epsilon)
    {
        return Enumerable.Repeat(0.5, values.Count).ToList();
    }
    return values.Select(v => (v - min) / (max - min)).ToList();
}

// Standard score (z-score) of a value within a sample.
public static double ZScore(double value, List<double> sample)
{
    double mean = sample.Average();
    double variance = sample.Sum(v => (v - mean) * (v - mean)) / sample.Count;
    double stdDev = Math.Sqrt(variance);
    if (stdDev == 0)
    {
        return 0;
    }
    return (value - mean) / stdDev;
}

// Simple moving average with a fixed window size.
public static List<double> MovingAverage(List<double> values, int window)
{
    var averages = new List<double>();
    double running = 0;
    for (int i = 0; i < values.Count; i++)
    {
        running += values[i];
        if (i >= window)
        {
            running -= values[i - window];
        }
        if (i >= window - 1)
        {
            averages.Add(running / window);
        }
    }
    return averages;
}

// Splits one CSV line into fields, honoring double-quoted cells.
public static List<string> ParseCsvLine(string line)
{
    var fields = new List<string>();
    var current = new System.Text.StringBuilder();
    bool inQuotes = false;
    for (int i = 0; i < line.Length; i++)
    {
        char c = line[i];
        if (inQuotes && c == '"' && i + 1 < line.Length && line[i + 1] == '"')
        {
            current.Append('"');
            i++;
        }
        else if (c == '"')
        {
            inQuotes = !inQuotes;
        }
        else if (c == ',' && !inQuotes)
        {
            fields.Add(current.ToString());
            current.Clear();
        }
        else
        {
            current.Append(c);
        }
    }
    fields.Add(current.ToString());
    return fields;
}

// Parses "key=value" pairs separated by newlines into a map.
public static Dictionary<string, string> ParseKeyValuePairs(string raw)
{
    var pairs = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
    foreach (string line in raw.Split('\n'))
    {
        int eq = line.IndexOf('=');
        if (eq <= 0)
        {
            continue;
        }
        string key = line[..eq].Trim();
        string value = line[(eq + 1)..].Trim();
        pairs[key] = value;
    }
    return pairs;
}

// Tallies log-level keywords found in a list of log lines.
public static Dictionary<string, int> SummarizeLogLevels(List<string> lines)
{
    string[] levels = { "DEBUG", "INFO", "WARN", "ERROR" };
    var counts = new Dictionary<string, int>();
    foreach (string line in lines)
    {
        string upper = line.ToUpperInvariant();
        string? hit = levels.FirstOrDefault(level => upper.Contains(level, StringComparison.Ordinal));
        if (hit != null)
        {
            counts[hit] = counts.GetValueOrDefault(hit) + 1;
        }
    }
    return counts;
}

// Removes records that share the same id, keeping the first one.
public static List<Dictionary<string, string>> DedupeByKey(
    List<Dictionary<string, string>> rows, string key)
{
    var seen = new HashSet<string>();
    var unique = new List<Dictionary<string, string>>();
    foreach (var row in rows)
    {
        if (seen.Add(row.GetValueOrDefault(key) ?? ""))
        {
            unique.Add(row);
        }
    }
    return unique;
}

// Sorts rows first by category, then by amount descending.
public static List<Dictionary<string, object>> SortByMultipleCriteria(
    List<Dictionary<string, object>> rows)
{
    return rows
        .OrderBy(r => (string)r["category"])
        .ThenByDescending(r => (double)r["amount"])
        .ToList();
}

// Buckets a list of numbers into ranges of ten, e.g. 10s, 20s.
public static Dictionary<int, List<int>> BucketByTens(List<int> values)
{
    return values
        .GroupBy(v => v / 10 * 10)
        .OrderBy(g => g.Key)
        .ToDictionary(g => g.Key, g => g.ToList());
}

// Running totals: result[i] = sum of values[0..i].
public static List<int> CumulativeSum(List<int> values)
{
    var totals = new List<int>(values.Count);
    int running = 0;
    foreach (int value in values)
    {
        running += value;
        totals.Add(running);
    }
    return totals;
}

// Drops statistical outliers using the interquartile range rule.
public static List<double> RemoveOutliers(List<double> values)
{
    var sorted = values.OrderBy(v => v).ToList();
    double q1 = sorted[sorted.Count / 4];
    double q3 = sorted[sorted.Count * 3 / 4];
    double iqr = q3 - q1;
    double lower = q1 - 1.5 * iqr;
    double upper = q3 + 1.5 * iqr;
    return values.Where(v => v >= lower && v <= upper).ToList();
}

// Transposes a rectangular matrix of doubles.
public static double[,] TransposeMatrix(double[,] matrix)
{
    int rows = matrix.GetLength(0);
    int cols = matrix.GetLength(1);
    var transposed = new double[cols, rows];
    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < cols; j++)
        {
            transposed[j, i] = matrix[i, j];
        }
    }
    return transposed;
}

// Dot product of two equal-length vectors.
public static double DotProduct(double[] a, double[] b)
{
    if (a.Length != b.Length)
    {
        throw new ArgumentException("vectors must have equal length");
    }
    return a.Zip(b, (x, y) => x * y).Sum();
}

// Counts how many records fall into each named category.
public static Dictionary<string, int> CountByCategory(List<string> categories)
{
    return categories.GroupBy(c => c).ToDictionary(g => g.Key, g => g.Count());
}

// Builds display rows, formatting each field for a fixed-width table.
public static List<string> BuildReportRows(
    List<Dictionary<string, object>> records, List<string> columns)
{
    var rows = new List<string>();
    foreach (var record in records)
    {
        var cells = columns.Select(column =>
            record.TryGetValue(column, out object? value)
                ? value is double d ? d.ToString("F2") : value.ToString() ?? "N/A"
                : "N/A");
        rows.Add(string.Join(" | ", cells));
    }
    return rows;
}

// Joins rows from two lists on a shared id column.
public static List<Dictionary<string, object>> JoinOnId(
    List<Dictionary<string, object>> left, List<Dictionary<string, object>> right)
{
    var joined = new List<Dictionary<string, object>>();
    foreach (var l in left)
    {
        foreach (var r in right)
        {
            if (Equals(l["id"], r["id"]))
            {
                var merged = new Dictionary<string, object>(l);
                foreach ((string key, object value) in r)
                {
                    merged[key] = value;
                }
                joined.Add(merged);
            }
        }
    }
    return joined;
}

// Downsamples a series by keeping every nth element.
public static List<double> Downsample(List<double> values, int step)
{
    return values.Where((_, index) => index % step == 0).ToList();
}

// Pivots a flat list of (category, status) pairs into counts per status.
public static Dictionary<string, Dictionary<string, int>> PivotStatusCounts(
    List<string> categories, List<string> statuses)
{
    var pivot = new Dictionary<string, Dictionary<string, int>>();
    for (int i = 0; i < categories.Count; i++)
    {
        if (!pivot.TryGetValue(categories[i], out Dictionary<string, int>? statusCounts))
        {
            statusCounts = new Dictionary<string, int>();
            pivot[categories[i]] = statusCounts;
        }
        statusCounts[statuses[i]] = statusCounts.GetValueOrDefault(statuses[i]) + 1;
    }
    return pivot;
}

// Trims and collapses whitespace on every string field of a record.
public static Dictionary<string, string> SanitizeRecord(Dictionary<string, string> record)
{
    return record.ToDictionary(
        kv => kv.Key,
        kv => string.IsNullOrWhiteSpace(kv.Value)
            ? ""
            : System.Text.RegularExpressions.Regex.Replace(kv.Value.Trim(), "\s+", " "));
}

File: dictionary/lua/data_processing.lua
-- mean returns the arithmetic average of a list of numbers.
local function mean(values)
	local total = 0
	for _, v in ipairs(values) do
		total = total + v
	end
	return total / #values
end

-- stddev returns the population standard deviation.
local function stddev(values)
	local m = mean(values)
	local total = 0
	for _, v in ipairs(values) do
		total = total + (v - m) ^ 2
	end
	return math.sqrt(total / #values)
end

-- clamp confines a value to the inclusive range [lo, hi].
local function clamp(value, lo, hi)
	if value < lo then
		return lo
	end
	if value > hi then
		return hi
	end
	return value
end

-- parse_csv_line splits a CSV row, honouring quoted fields.
local function parse_csv_line(line)
	local fields = {}
	local current = ""
	local in_quotes = false
	local i = 1
	while i <= #line do
		local ch = line:sub(i, i)
		if ch == '"' then
			if in_quotes and line:sub(i + 1, i + 1) == '"' then
				current = current .. '"'
				i = i + 1
			else
				in_quotes = not in_quotes
			end
		elseif ch == "," and not in_quotes then
			fields[#fields + 1] = current
			current = ""
		else
			current = current .. ch
		end
		i = i + 1
	end
	fields[#fields + 1] = current
	return fields
end

-- parse_int_list converts "1,2,3" into a list of numbers.
local function parse_int_list(s)
	local result = {}
	for part in s:gmatch("[^,]+") do
		local value = tonumber(part:match("^%s*(.-)%s*$"))
		if not value then
			return nil
		end
		result[#result + 1] = value
	end
	return result
end

-- filter_outliers drops values more than 3 stddevs from the mean.
local function filter_outliers(values)
	if #values < 3 then
		return values
	end
	local m = mean(values)
	local sd = stddev(values)
	if sd == 0 then
		return values
	end
	local result = {}
	for _, v in ipairs(values) do
		if math.abs(v - m) <= 3 * sd then
			result[#result + 1] = v
		end
	end
	return result
end

-- normalize_values scales a list into the [0, 1] range.
local function normalize_values(values)
	if #values == 0 then
		return values
	end
	local lo, hi = values[1], values[1]
	for _, v in ipairs(values) do
		lo = math.min(lo, v)
		hi = math.max(hi, v)
	end
	local span = hi - lo
	local result = {}
	if span == 0 then
		for i = 1, #values do
			result[i] = 0
		end
		return result
	end
	for i, v in ipairs(values) do
		result[i] = (v - lo) / span
	end
	return result
end

-- zscore transforms values into standard scores.
local function zscore(values)
	local m = mean(values)
	local sd = stddev(values)
	local result = {}
	if sd == 0 then
		for i = 1, #values do
			result[i] = 0
		end
		return result
	end
	for i, v in ipairs(values) do
		result[i] = (v - m) / sd
	end
	return result
end

-- histogram counts occurrences per exact value.
local function histogram(values)
	local counts = {}
	for _, v in ipairs(values) do
		counts[v] = (counts[v] or 0) + 1
	end
	return counts
end

-- cumulative_sum returns the running total at each index.
local function cumulative_sum(values)
	local result = {}
	local total = 0
	for i, v in ipairs(values) do
		total = total + v
		result[i] = total
	end
	return result
end

-- moving_average smooths a series with a sliding window.
local function moving_average(values, window)
	local result = {}
	local total = 0
	for i = 1, #values do
		total = total + values[i]
		if i > window then
			total = total - values[i - window]
		end
		result[i] = total / math.min(i, window)
	end
	return result
end

-- delta computes the difference between consecutive values.
local function delta(values)
	if #values < 2 then
		return {}
	end
	local result = {}
	for i = 2, #values do
		result[i - 1] = values[i] - values[i - 1]
	end
	return result
end

-- clamp_values restricts every value to [lo, hi] in place.
local function clamp_values(values, lo, hi)
	for i, v in ipairs(values) do
		values[i] = clamp(v, lo, hi)
	end
	return values
end

-- dedupe_preserve_order removes repeats keeping first-seen order.
local function dedupe_preserve_order(values)
	local seen = {}
	local result = {}
	for _, v in ipairs(values) do
		if not seen[v] then
			seen[v] = true
			result[#result + 1] = v
		end
	end
	return result
end

-- partition splits values into two lists by a predicate.
local function partition(values, predicate)
	local trues, falses = {}, {}
	for _, v in ipairs(values) do
		if predicate(v) then
			trues[#trues + 1] = v
		else
			falses[#falses + 1] = v
		end
	end
	return trues, falses
end

-- min_max_scale normalises values between new_min and new_max.
local function min_max_scale(values, new_min, new_max)
	if #values == 0 then
		return values
	end
	local lo, hi = values[1], values[1]
	for _, v in ipairs(values) do
		lo = math.min(lo, v)
		hi = math.max(hi, v)
	end
	local result = {}
	if hi == lo then
		for i = 1, #values do
			result[i] = (new_min + new_max) / 2
		end
		return result
	end
	for i, v in ipairs(values) do
		result[i] = new_min + (v - lo) / (hi - lo) * (new_max - new_min)
	end
	return result
end

-- detect_peaks finds indices that exceed both neighbours.
local function detect_peaks(values)
	local peaks = {}
	for i = 2, #values - 1 do
		if values[i] > values[i - 1] and values[i] > values[i + 1] then
			peaks[#peaks + 1] = i
		end
	end
	return peaks
end

-- group_consecutive groups equal values into {value, count} pairs.
local function group_consecutive(values)
	if #values == 0 then
		return {}
	end
	local runs = {}
	local current = values[1]
	local count = 1
	for i = 2, #values do
		if values[i] == current then
			count = count + 1
		else
			runs[#runs + 1] = { value = current, count = count }
			current = values[i]
			count = 1
		end
	end
	runs[#runs + 1] = { value = current, count = count }
	return runs
end

-- running_average computes the cumulative average at each step.
local function running_average(values)
	local result = {}
	local total = 0
	for i, v in ipairs(values) do
		total = total + v
		result[i] = total / i
	end
	return result
end

-- pairwise_add combines two same-length lists element-wise.
local function pairwise_add(a, b)
	if #a ~= #b then
		return nil
	end
	local result = {}
	for i = 1, #a do
		result[i] = a[i] + b[i]
	end
	return result
end

-- interleave merges two lists alternating their elements.
local function interleave(a, b)
	local result = {}
	local n = math.max(#a, #b)
	for i = 1, n do
		if a[i] ~= nil then
			result[#result + 1] = a[i]
		end
		if b[i] ~= nil then
			result[#result + 1] = b[i]
		end
	end
	return result
end

-- windowed pulls every contiguous window of size k.
local function windowed(values, size)
	if size <= 0 or size > #values then
		return {}
	end
	local result = {}
	for start = 1, #values - size + 1 do
		local window = {}
		for i = 0, size - 1 do
			window[i + 1] = values[start + i]
		end
		result[#result + 1] = window
	end
	return result
end

-- bucketize maps values into n equal-width buckets and counts them.
local function bucketize(values, buckets)
	if buckets <= 0 or #values == 0 then
		return {}
	end
	local lo, hi = values[1], values[1]
	for _, v in ipairs(values) do
		lo = math.min(lo, v)
		hi = math.max(hi, v)
	end
	local counts = {}
	for i = 1, buckets do
		counts[i] = 0
	end
	local width = (hi - lo) / buckets
	for _, v in ipairs(values) do
		local index = math.floor((v - lo) / width) + 1
		index = clamp(index, 1, buckets)
		counts[index] = counts[index] + 1
	end
	return counts
end

-- cross_correlation measures similarity of two same-length series.
local function cross_correlation(a, b)
	if #a ~= #b or #a == 0 then
		return 0
	end
	local ma, mb = mean(a), mean(b)
	local num, da, db = 0, 0, 0
	for i = 1, #a do
		local x = a[i] - ma
		local y = b[i] - mb
		num = num + x * y
		da = da + x * x
		db = db + y * y
	end
	local denom = math.sqrt(da * db)
	if denom == 0 then
		return 0
	end
	return num / denom
end
